crates/filter/src/source.rs

```diff
@@ -61,18 +68,22 @@ impl<I: Iterator<Item = String>> Source<I> {
                 .lines()
                 .filter_map(|lines_iter| {
                     lines_iter.ok().and_then(|line| {
-                        matcher(&line).map(|(score, indices)| (line, score, indices))
+                        matcher(&line).map(|(score, indices)| (SourceItem::new(line), score, indices))
                     })
                 })
                 .collect::<Vec<_>>(),
             Self::File(fpath) => std::fs::read_to_string(fpath)?
                 .par_lines()
                 .filter_map(|line| {
-                    matcher(&line).map(|(score, indices)| (line.into(), score, indices))
+                    matcher(&line).map(|(score, indices)| (SourceItem::new(line.into()), score, indices))
                 })
                 .collect::<Vec<_>>(),
             Self::List(list) => list
-                .filter_map(|line| matcher(&line).map(|(score, indices)| (line, score, indices)))
+                .filter_map(|item| {
+                    // XXX: How do I avoid cloning here?
+                    let line = item.filter.clone().unwrap_or(item.display.clone());
+                    matcher(&line).map(|(score, indices)| (item, score, indices))
```

Comment:
You probably missed that one in the `unwrap_or` function docs, so here it is: whenever you do something heavy in `_or` part, use `unwrap_or_else(|| /* heavy part */ )`. Right now you clone `display` part on *every* item, and then, if `filter` part is `Some`, it is dropped.

```rust
matcher(item.filter.as_deref().unwrap_or(&item.display)).map(|(score, indices)| (item, score, indices))
```

Reply:
Yes I had guessed but I needed your input on the other cloning thing so I did the simplest thing I could. Seems you often solve my cloning issues with `.as_deref()`, I'll read on that.

Reply:
There's no magic, `as_deref` is just a function with `match &x { Some(x) => Some(x.deref()), None => None }` but writing `as_deref` is just simpler than writing `match`. Same goes for `as_ref()`, `as_mut` and [other `Option` methods](https://doc.rust-lang.org/std/option/enum.Option.html) — they all could be implemented with `match`, no magic.